nearform/titus
# feat: authorization with casbin

Closes #657 

Adds authorization with [fastify-casbin](https://github.com/nearform/fastify-casbin).

A few things to consider:

- Titus does not have routes, intentionally. The only routes used for other purposes are the ones  when cognito is used as an authentication mechanism, which load the users and the current user from cognito and display them in the (only) page of the application
- Therefore, I used those routes to apply a very simple authorization rule which is: if you belong to the cognito `admin` group you can delete users. Note that an admin group is something that you would set up yourself in the cognito user pool, it's not an built-in concept

**I have highlighted the relevant changes in the code, because most of the changes are just enablers for this feature.**

In order to test it, you need to have access to a cognito user pool, setting up which is not straightforward, therefore please reach out to me and I can provide you with IAM credentials to access my user pool.

## See it in action

In the image you can see how it works. Initially, the user is not in the admin group, therefore it cannot delete any users.
Then, it is added to the admin group and it can delete the other user.

![Untitled_ Nov 6, 2020 6_35 PM (1)](https://user-images.githubusercontent.com/20181/98397923-6a760200-2060-11eb-8a6f-1839f6ee3083.gif)

## How to test

- `lerna bootstrap`
- add the necessary AWS/Cognito information to the .env files of both frontend and backend (reach out to me to get the env var values)
- configure a local AWS profile for accessing Cognito with IAM credentials (reach out to me)
- `npm run start:all`
- login with a user and see the behavior based on whether the user is part of the admin group or not
